### ISSUE TITLE
After registration, is the order of transforms composition specified in the documentation correct?

### ISSUE THREAD

**Issue by sdiebolt** [SimpleITK/SimpleITK#2043]
**Describe the bug**

The order of transform composition after registration might be wrong in most of the documentation and tutorial notebooks.

**To Reproduce**

In the `05_basic_registration.ipynb` notebook, compare the resampling that uses the composite transform created by the current version of the notebook, i.e.

```python
final_transform_v4 = sitk.CompositeTransform(
    [registration_method.Execute(fixed_image, moving_image), initial_transform]
)
```

![image](https://github.com/SimpleITK/SimpleITK/assets/3774850/b038bd69-9d87-4eaf-abea-4da668a7e113)

with the composite transform that uses the correct order (following ITK's book):

```python
final_transform_v4 = sitk.CompositeTransform(
    [initial_transform, registration_method.Execute(fixed_image, moving_image)]
)
```

![image](https://github.com/SimpleITK/SimpleITK/assets/3774850/d35548cd-5cca-48fd-b708-0d5a81ae8298)

The second resampling seems better to me than the first.

To see the error even more clearly, we can use a moving initial transform that performs a flip along one axis. For example, add the following to the 7th code cell before executing the registration:

```python
initialization = np.eye(4)
initialization[0, 0] = -1
initialization[0, 3] = moving_image.GetSize()[0] * moving_image.GetSpacing()[0]

moving_initial_transform = sitk.AffineTransform(
    initialization[:3, :3].flatten(), initialization[:3, 3], (0,) * 3
)
registration_method.SetMovingInitialTransform(moving_initial_transform)
```

The registration should still be satisfactory due to the symmetry of the brain image.

Composing the optimized and moving initial transforms as described in [SimpleITK's documentation](https://simpleitk.readthedocs.io/en/release/registrationOverview.html#transforms-and-image-spaces) leads to an incorrect result:

![image](https://github.com/SimpleITK/SimpleITK/assets/3774850/ecd777d7-cd17-41a8-814f-98bda0bff0d0)

However, composing these transforms using

```python
composite_transform = sitk.CompositeTransform([moving_initial_transform, final_transform])
```

as is described in [section 3.2 of ITK's book](https://itk.org/ITKSoftwareGuide/html/Book2/ITKSoftwareGuide-Book2ch3.html#x26-1000003.2) leads to the correct result:

![image](https://github.com/SimpleITK/SimpleITK/assets/3774850/fd3fb50d-73c4-49ac-8882-ec3c28577401)

1. Operating system, version, and architecture
 - OS: Ubuntu 22.04.3 LTS
 - Architecture: x86_64
2. Programming language, and version: Python 3.11.
3. Version of SimpleITK: 2.3.1
4. How was SimpleITK installed? pip
5. A minimal working example which causes the error. See above.

**Expected behavior**

The resampling should be accurate using the moving initial transforms above.

**Additional context**

I originally opened a [topic on ITK's Discourse](https://discourse.itk.org/t/how-do-you-compose-transforms-after-registration-to-perform-resampling/6401) as I was confused about the discrepancies between SimpleITK's documentation and the resampling results I got when using axis flips in the moving initial transform. After further research, I believe there is a discrepancy between ITK's documentation on how to compose transforms after registration and SimpleITK's. Notably, most of the public GitHub repositories that use SimpleITK already apply the correct composition order (i.e. the inverse order of what the documentation specifies). The [`ImageRegistrationMethodDisplacement1.py` example from SimpleITK](https://github.com/SimpleITK/SimpleITK/blob/master/Examples/ImageRegistrationMethodDisplacement1/ImageRegistrationMethodDisplacement1.py) is also using the correct order (sparking two questions on Discourse, [here](https://discourse.itk.org/t/is-order-of-transforms-in-a-composite-transform-reversed-relative-to-transform-order-during-registration/3829) and [here](https://discourse.itk.org/t/image-registration-method-displacement-1-example-composite-transform-order/3920), that got no explicit answers).

Actually, I think the error is illustrated by this figure from one of SimpleITK's courses:

![image](https://github.com/SimpleITK/SimpleITK/assets/3774850/3877631b-05b5-4675-bd7a-5ce1ca9cabad)

Shouldn't $T_\text{opt}$ actually go from virtual image space to virtual image space, since the registration is performed in the virtual space? Then the final output transform would be $T_\text{m}\circ T_\text{opt}\circ T_\text{f}^{-1}$ like described in ITK's guide, and not $T_\text{opt}\circ T_\text{m}\circ T_\text{f}^{-1}$ like in SimpleITK's documentation.

I hope I'm wrong however, and would very much appreciate your help in understanding this problem!

**Comment by sdiebolt** [comment]
After further testing, it may be that the order of transforms composition described in SimpleITK's documentation and example notebooks is wrong in the general case. Compared to ITK's documentation, the moving initial transform and optimized transforms may be composed in the wrong order.

Here's the registration result given by the 8th code cell of example notebook `05_basic_registration.ipynb`:

![image](https://github.com/SimpleITK/SimpleITK/assets/3774850/04463158-f4c7-45c7-9e9b-994736ce6a44)

And here's the result obtained by composing the transforms following ITK's documentation, i.e.

```python
final_transform_v4 = sitk.CompositeTransform(
    [initial_transform, registration_method.Execute(fixed_image, moving_image)]
)
```

![image](https://github.com/SimpleITK/SimpleITK/assets/3774850/77c19ab3-1e63-4525-be5c-f0dea0b28837)

The second resampling seems better to me than the first. And that's without modifying the moving initial transform.

Actually, I think the error is illustrated by this figure from one of SimpleITK's courses:

![image](https://github.com/SimpleITK/SimpleITK/assets/3774850/3877631b-05b5-4675-bd7a-5ce1ca9cabad)

Shouldn't $T_\text{opt}$ actually be inside the virtual image space? Then the final output transform would be $T_\text{m}\circ T_\text{opt}\circ T_\text{f}^{-1}$ like described in ITK's guide, and not $T_\text{opt}\circ T_\text{m}\circ T_\text{f}^{-1}$ like in SimpleITK's documentation.

I hope I'm wrong however, and would very much appreciate your help in understanding this problem!

**Comment by blowekamp** [comment]
Thank you for pointing this out. It looks like the [Registration Overview](https://github.com/SimpleITK/SimpleITK/blob/master/docs/source/registrationOverview.rst) documentation in this repository needs to be updated. 

Another issue has also been created in the Notebooks repository. Please let us know if you have encountered other places (that can be updated) which container this wrong order of transformations.

**Comment by sdiebolt** [comment]
I've found this error in two main places: the "Registration Overview" section of the documentation that you already pointed out, and all versions of the [`XX_basic_registration.ipynb`](https://github.com/SimpleITK/SPIE2019_COURSE/blob/98e01266ebc6416814b6da194ee697fbbddf7631/04_basic_registration.ipynb#L316) notebook (the error in present in both R and Python notebooks for all such courses, i.e. SPIE2018, ISBI2020, the [SimpleITK TUTORIAL](https://github.com/SimpleITK/TUTORIAL/blob/main/05_basic_registration.ipynb) repository, etc.).

I think most of these notebooks are in archived repositories, but since they often come up highly ranked in e.g. Google searches (the SPIE2019 notebook comes up as one of the top results when looking for "SimpleITK registration" on a fresh browser), maybe they should still be updated?

I'd be happy to open PRs to fix these issues if that's of any help!

**Comment by blowekamp** [comment]
Thank you for offering to contribute. Identifying the locations of the places that need to be updated is a very important and critical step.

For this repository, we are currently looking for an original copy of the figure used in the RtD documentation to update the transformation figure and the the formula. @zivy

**Comment by zivy** [comment]
Addressed by PR #2055
